modules/backup-azure/client.go

```diff
@@ -0,0 +1,177 @@
+	config     clientConfig
+	serviceURL string
+	dataPath   string
+}
+
+func newClient(ctx context.Context, config *clientConfig, dataPath string) (*azureClient, error) {
+	connectionString := os.Getenv("AZURE_STORAGE_CONNECTION_STRING")
+	if connectionString != "" {
+		client, err := azblob.NewClientFromConnectionString(connectionString, nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "create client using connection string")
+		}
+		return &azureClient{client, *config, "", dataPath}, nil
+	}
+
+	// Your account name and key can be obtained from the Azure Portal.
+	accountName := os.Getenv("AZURE_STORAGE_ACCOUNT_NAME")
+	accountKey := os.Getenv("AZURE_STORAGE_PRIMARY_ACCOUNT_KEY")
+
+	if accountName == "" {
+		return nil, errors.New("AZURE_STORAGE_ACCOUNT_NAME must be set")
+	}
+
+	// The service URL for blob endpoints is usually in the form: http(s)://<account>.blob.core.windows.net/
+	serviceURL := fmt.Sprintf("https://%s.blob.core.windows.net/", accountName)
+
+	if accountKey != "" {
+		cred, err := azblob.NewSharedKeyCredential(accountName, accountKey)
+		if err != nil {
+			return nil, err
+		}
+
+		client, err := azblob.NewClientWithSharedKeyCredential(serviceURL, cred, nil)
+		if err != nil {
+			return nil, err
+		}
+		return &azureClient{client, *config, serviceURL, dataPath}, nil
+	}
+
+	client, err := azblob.NewClientWithNoCredential(serviceURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &azureClient{client, *config, serviceURL, dataPath}, nil
+}
+
+func (a *azureClient) HomeDir(backupID string) string {
+	return a.serviceURL + path.Join(a.config.Container, a.makeObjectName(backupID))
+}
+
+func (a *azureClient) makeObjectName(parts ...string) string {
+	base := path.Join(parts...)
+	return path.Join(a.config.BackupPath, base)
+}
+
+func (a *azureClient) GetObject(ctx context.Context, backupID, key string) ([]byte, error) {
+	objectName := a.makeObjectName(backupID, key)
+
+	if err := ctx.Err(); err != nil {
+		return nil, backup.NewErrContextExpired(errors.Wrapf(err, "get object '%s'", objectName))
+	}
+
+	blobDownloadResponse, err := a.client.DownloadStream(ctx, a.config.Container, objectName, nil)
+	if err != nil {
+		if bloberror.HasCode(err, bloberror.BlobNotFound) {
+			return nil, backup.NewErrNotFound(errors.Wrapf(err, "get object '%s'", objectName))
+		}
+		return nil, backup.NewErrInternal(errors.Wrapf(err, "download stream for object '%s'", objectName))
+	}
+
+	reader := blobDownloadResponse.Body
+	downloadData, err := io.ReadAll(reader)
+	if err != nil {
+		return nil, backup.NewErrInternal(errors.Wrapf(err, "read stream for object '%s'", objectName))
+	}
+	err = reader.Close()
+	if err != nil {
+		return nil, backup.NewErrInternal(errors.Wrapf(err, "close stream for object '%s'", objectName))
+	}
+
+	return downloadData, nil
+}
+
+func (a *azureClient) PutFile(ctx context.Context, backupID, key, srcPath string) error {
+	srcPath = path.Join(a.dataPath, srcPath)
+	contents, err := os.ReadFile(srcPath)
```

Comment:
There was an issue with the GCS backup the other day where large files were causing problems when the contents were first loaded into memory, and then written to storage.

To solve this, the file handle was passed to the GCS client, and the contents were written directly to storage without being read into memory first. 

Do we have an option like here that for Azure?

Reply:
Yes, I have addressed that, I have added here a method that uses file handlers. Thanks for bringing that up

Reply:
Nice 👍